### DISCUSSION TITLE
napi_add_env_cleanup_hook when process exits forcefully

### DISCUSSION THREAD

**Discussion by kwonoj** [nodejs/help#4917] [Q&A]
relative context: https://github.com/vercel/next.js/pull/35803#discussion_r842021435=

There is a codebase I'm using `napi_add_env_cleanup_hook` to initiate teardown when a node.js process lifecycle completes. It was working as expected under normal circumstances, but it looks like it is not being called when process exits explitly: roughly,

// working
```
initSome() //internally calls napi_add_env_cleanup_hook
..
// cleanup_hook called when node.js exits gracefully
```

// not working
```
initSome() //internally calls napi_add_env_cleanup_hook
..
process.exit() // cleanup_hook is not being called
```

Peeking process.exit() doc for its behavior, it _looks like_ expected behavior

> Calling process.exit() will force the process to exit as quickly as possible even if there are still asynchronous operations pending that have not yet completed fully, including I/O operations to process.stdout and process.stderr.

as it does not give any gaurantee to handling / flushing out remaining task in the queue.

- Am I reading correct, that this is an expected behavior?
- If that's the case, is there way to trigger `napi_add_env_cleanup_hook` in a graceful way regardless of how process exits?

I've managed to expose a new interface for explicit teardown and chaning it via `process.on('exit')`. It seems to work, but I hope if there's simplified way works automatically across all exits with automatic cleaning hook.

**Comment by mhdawson** [comment]
@gabrielschulhof can correct me if I'm wrong but I think what you say is correct. process.exit() is exit without cleanup and if asynchronous operations are not complete it won't be safe to tear down an env and run the cleanup hooks in general.
